### PR TITLE
Event images

### PR DESCRIPTION
**Pull request by mic4ael**

### CODE REVIEW

File: indico/modules/events/layout/controllers/images.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def _process(self):
                 # Invalid image data
                 continue
             data.seek(0)
-            if image_type not in {'jpeg', 'gif', 'png'}:
+            if image_type not in {'jpg', 'jpeg', 'gif', 'png'}:
```
Comment (ThiefMaster) [inline indico/modules/events/layout/controllers/images.py]: does this ever happen?

File: indico/modules/events/layout/controllers/layout.py [hunk 2]
```diff
@@ -17,8 +17,9 @@
 from __future__ import unicode_literals
 
 import binascii
-import mimetypes
 from io import BytesIO
+from os.path import splitext
+from PIL import Image
```
Comment (ThiefMaster) [inline indico/modules/events/layout/controllers/layout.py]: that's not stdlib

File: indico/modules/events/layout/controllers/layout.py [hunk 3]
```diff
@@ -17,8 +17,9 @@
 from __future__ import unicode_literals
 
 import binascii
-import mimetypes
 from io import BytesIO
+from os.path import splitext
```
Comment (ThiefMaster) [inline indico/modules/events/layout/controllers/layout.py]: usually os/os.path stuff is called with the full name (`os.path.splitext`)

File: indico_zodbimport/modules/event_layout.py [hunk 4]
```diff
@@ -83,12 +86,23 @@ def migrate_layout_settings(self):
                                      event_id=event.id)
                     continue
                 with open(path, 'rb') as f:
-                    logo_content = f.read()
-                logo_content_type = mimetypes.guess_type(logo.fileName)[0]
+                    try:
+                        logo_image = Image.open(f)
+                        logo_content_type = 'image/' + logo_image.format.lower()
```
Comment (ThiefMaster) [inline indico_zodbimport/modules/event_layout.py]: why? you are converting it to png

File: indico_zodbimport/modules/event_layout.py [hunk 5]
```diff
@@ -83,12 +86,23 @@ def migrate_layout_settings(self):
...
+                        logo_bytes = BytesIO()
+                        logo_image.save(logo_bytes, 'PNG')
+                        logo_bytes.seek(0)
+                        logo_content = logo_bytes.read()
+                        logo_filename = splitext(secure_filename(convert_to_unicode(logo.fileName), 'logo'))[0] + '.png'
+                    except IOError:
+                        self.print_warning("Can't open the event logo file. Probably not an image?")
```
Comment (ThiefMaster) [inline indico_zodbimport/modules/event_layout.py]: skip it in this case

File: indico/modules/events/layout/controllers/layout.py [hunk 6]
```diff
@@ -95,14 +96,22 @@ def _process(self):
 class RHLayoutLogoUpload(RHLayoutBase):
     def _process(self):
         f = request.files['file']
-        content = f.read()
+        try:
+            img = Image.open(f)
+        except IOError:
+            return jsonify_data(content={})
+        if img.format.lower() not in {'jpg', 'jpeg', 'png', 'gif'}:
```
Comment (ThiefMaster) [inline indico/modules/events/layout/controllers/layout.py]: maybe we could flash() an error message here?

File: indico/modules/events/layout/controllers/images.py (hunk 1 above)
Comment (mic4ael) [inline indico/modules/events/layout/controllers/images.py]: Yeah, that happened to me a few hours ago. That is why I have added it here.

File: indico/modules/events/layout/controllers/layout.py [hunk 7]
```diff
@@ -95,14 +96,24 @@ def _process(self):
 class RHLayoutLogoUpload(RHLayoutBase):
     def _process(self):
         f = request.files['file']
-        content = f.read()
+        try:
+            img = Image.open(f)
+        except IOError:
+            flash(_("You can't upload this file as a logo."))
```
Comment (ThiefMaster) [inline indico/modules/events/layout/controllers/layout.py]: "cannot" and `'error'` as the second argument

File: indico/modules/events/layout/controllers/layout.py [hunk 8]
```diff
@@ -95,14 +96,24 @@ def _process(self):
...
+            return jsonify_data(content={})
+        if img.format.lower() not in {'jpg', 'jpeg', 'png', 'gif'}:
```
Comment (ThiefMaster) [inline indico/modules/events/layout/controllers/layout.py]: I still it's better to figure out how you ended up with 'jpg' here and if you can't reproduce it remove it again...

File: indico/modules/events/layout/controllers/layout.py (hunk 7 above)
Comment (mic4ael) [inline indico/modules/events/layout/controllers/layout.py]: Ah, thanks. Forgot about the second argument.

File: indico_zodbimport/modules/event_layout.py [hunk 9]
```diff
@@ -83,12 +86,23 @@ def migrate_layout_settings(self):
...
+                        logo_image = Image.open(f)
+                        logo_content_type = 'image/png'
+                        logo_bytes = BytesIO()
+                        logo_image.save(logo_bytes, 'PNG')
+                        logo_bytes.seek(0)
+                        logo_content = logo_bytes.read()
+                        logo_filename = (os.path.splitext(secure_filename(convert_to_unicode(logo.fileName),
+                                                                          'logo'))[0] + '.png')
+                    except IOError:
+                        logo_content = f.read()
```
Comment (ThiefMaster) [inline indico_zodbimport/modules/event_layout.py]: IMHO we should simply skip "broken" logos

File: indico_zodbimport/modules/event_layout.py [hunk 10]
```diff
@@ -83,12 +86,23 @@ def migrate_layout_settings(self):
...
-                    logo_content = f.read()
-                logo_content_type = mimetypes.guess_type(logo.fileName)[0]
+                    try:
+                        logo_image = Image.open(f)
+                        logo_content_type = 'image/png'
+                        logo_bytes = BytesIO()
+                        logo_image.save(logo_bytes, 'PNG')
+                        logo_bytes.seek(0)
+                        logo_content = logo_bytes.read()
+                        logo_filename = (os.path.splitext(secure_filename(convert_to_unicode(logo.fileName),
```
Comment (ThiefMaster) [inline indico_zodbimport/modules/event_layout.py]: that line is incredibly ugly

File: indico_zodbimport/modules/event_layout.py [hunk 11]
```diff
@@ -83,12 +86,23 @@ def migrate_layout_settings(self):
                                      event_id=event.id)
                     continue
                 with open(path, 'rb') as f:
-                    logo_content = f.read()
-                logo_content_type = mimetypes.guess_type(logo.fileName)[0]
+                    try:
+                        logo_image = Image.open(f)
```
Comment (ThiefMaster) [inline indico_zodbimport/modules/event_layout.py]: only this line should be in the `try` block (or does it ever fail somewhere else with an IOError?)

File: indico_zodbimport/modules/event_layout.py (hunk 9 above)
Comment (mic4ael) [inline indico_zodbimport/modules/event_layout.py]: That was my first approach. However I think we should leave it to the users to change it.

**Comment by ThiefMaster** [comment]
_bp_images.add_url_rule('/logo-<slug>', 'logo_display', RHLogoDisplay)

I'd add `.png` here

File: indico_zodbimport/modules/event_layout.py (hunk 11 above)
Comment (mic4ael) [inline indico_zodbimport/modules/event_layout.py]: Only this line, however from now on it is the only line wrapped with `try`.

File: indico/modules/events/layout/controllers/layout.py [hunk 12]
```diff
@@ -95,14 +96,24 @@ def _process(self):
...
     def _process(self):
         f = request.files['file']
-        content = f.read()
+        try:
+            img = Image.open(f)
+        except IOError:
+            flash(_('You cannot upload this file as a logo.', 'error'))
+            return jsonify_data(content={})
+        if img.format.lower() not in {'jpeg', 'png', 'gif'}:
+            flash(_("You can't upload file of this format."))
```
Comment (ThiefMaster) [inline indico/modules/events/layout/controllers/layout.py]: `'error'`

File: indico/modules/events/layout/controllers/layout.py [hunk 13]
```diff
@@ -95,14 +96,24 @@ def _process(self):
...
+        image_bytes.seek(0)
+        content = image_bytes.read()
         self.event.logo = content
-        content_type = mimetypes.guess_type(f.filename)[0] or f.mimetype or 'application/octet-stream'
         self.event.logo_metadata = {
             'hash': binascii.crc32(content) & 0xffffffff,
             'size': len(content),
-            'filename': f.filename,
-            'content_type': content_type
+            'filename': os.path.splitext(f.filename)[0] + '.png',
```
Comment (ThiefMaster) [inline indico/modules/events/layout/controllers/layout.py]: `secure_filename(f.filename, 'logo')`

File: indico/modules/events/layout/controllers/layout.py [hunk 14]
```diff
@@ -95,14 +96,24 @@ def _process(self):
...
     def _process(self):
         f = request.files['file']
-        content = f.read()
+        try:
+            img = Image.open(f)
+        except IOError:
+            flash(_('You cannot upload this file as a logo.', 'error'))
+            return jsonify_data(content={})
+        if img.format.lower() not in {'jpeg', 'png', 'gif'}:
+            flash(_('You cannot upload file of this format.', 'error'))
```
Comment (ThiefMaster) [inline indico/modules/events/layout/controllers/layout.py]: _("The file has an invalid format ({format}").format(format=img.format)

File: indico_zodbimport/modules/event_layout.py [hunk 15]
```diff
@@ -82,19 +84,28 @@ def migrate_layout_settings(self):
...
-                    'content_type': logo_content_type
-                }
-                e.logo = logo_content
-                e.logo_metadata = logo_metadata
-                if not self.quiet:
-                    self.print_success(cformat('- %{cyan}[Logo] {}').format(logo.fileName), event_id=event.id)
+                try:
+                    logo_image = Image.open(path)
+                except IOError:
+                    self.print_warning('Cannot open the file. The file probably is not an image.')
```
Comment (ThiefMaster) [inline indico_zodbimport/modules/event_layout.py]:
```
except IOError as e:
    "Cannot open {}: {}".format(path, e)
```

**Comment by ThiefMaster** [comment]
Merged bb92da1

**Pull request closed without merging**